### ISSUE TITLE
What about transient dependencies?

### ISSUE THREAD

**Issue by nicojs** [Romakita/ts-express-decorators#179]
## Informations

Type | Version
---|---
Question | 2.x

****

## Description

First of all, thanks for this awesome project! It is exactly the thing that I was looking for. A very elegant solution with great docs and even a focus on testing. 🥇 

I was wondering what your thoughts were about *transient* dependency injection. With transient, i mean dependencies that are created for each request.

With transient dependencies, i could do something like this:

```typescript
@Service()
class GithubAgent {
    private accessToken: string;
    constructor( @Req() request: Express.Request) {
        this.accessToken = request.user.accessToken;
    }

   getSomeResource() {
        return new SomeHttpClient(this.accessToken).get('https://api.github.com/some/resource');
   }
}

@Controller('/some/resource')
class SomeResourceController {

    constructor(private agent: GithubAgent) { }

    @Get('/')
    public get() {
       return this.agent.getSomeResource();
    }
}
```

Keep up the great work!

**Comment by lwallent** [comment]
Hi @nicojs,

Without knowing your exact demands I would say you should be looking at `@Middleware` and `@Locals`. Again it is all about testability and single responsibility principle.

I would never think to have the Express Request present in a Service.

So what I would suggest is that you set up a Middleware chain that extracts token or whatever you have and then further down the chain you would have a middleware injecting the user into the Response locals. Then you could have that user injected into your controller endpoint. This would also make it a lot easier to test your controller and service as they only have to know the user and not care about some Express Request object being passed around. 

I personally use a `@JWT` decorator on the controller where an authorised user is required:

```
export function JWT(): Function {
    return Store.decorate((store: Store) => {
        store
            .set(JWTMiddleware, {})
            .merge("operation", {
              security: [{"Token": []}],
              responses: {"401": {description: "No authorization token was found"}}
            });

        return UseBefore(JWTMiddleware, AddUserToLocalsMiddleware);
    });
} 
```
The AddUserToLocalsMiddleware will make sure that your user is present in the response locals.

```
@Middleware()
export class AddUserToLocalsMiddleware implements IMiddleware {

    constructor(private userService: UserService) {
    }

    async use(@Locals('token')          token:IJwtToken,          
              @Response()               response: Express.Response,
              @Next()                   next: Express.NextFunction) {

        if (!token) {
            next(new Forbidden('No token present'));
        }  
        
        let user = await this.userService.findByToken(token);
   
        // Response Locals is the official Express documented place 
        // for local variables scoped to the request, and therefore 
        // available only during that request/response cycle.
        if (user) {
            response.locals.user = user;
            next();
        } else {
            next(new BadRequest("The User is not present in the system"));
        }
    }
}
```

```
@Controller('/some/resource')
@JWT() 
class SomeResourceController {

    constructor(private agent: GithubAgent) { }

    @Get('/')
    public get(@Locals('user') user: IUser) {
       return this.agent.getSomeResource(user);
    }
}
```

These where  some quick pointers you may use. Let me know if this is helpful or you need any further.  


Kind Regards

Lasse

**Comment by Romakita** [comment]
Hi @nicojs ,

Thanks for your support :)
This subject has already been addressed and is in the backlog #58 . But, the feature require to use a Proxy JavaScript and it's little bit harder.  I need some help about this subject and a time also.

The temporary solution is to use the @Scope decorator (undocumented) on a controller like this:

```typescript
@Controller("/")
@Scope("request")
export MyController {
   constructor(private service){

   }
   @Get('/')
    public get() {
    }
}
```
> With Scope annotation, the controller will be instantiated for each incoming request. But the injected Services remain singletons.

I agree, this is not the same feature that you presented but maybe it'll help you :)

See you ;)
Romain

**Comment by Romakita** [comment]
@lwallent Awesome :)

**Comment by nicojs** [comment]
@Romakita thanks for pointing that out to me. That issue is exactly what I meant, I'll close this issue as it is a duplicate.

@lwallent thanks for your samples! I still would like to wait for @Romakita's fix of #58. Until then I will use a construct like this:

```typescript
  @Get('/:name')
  public get( @PathParams('name') name: string, @Req() request: Express.Request): Promise<Login> {
        return new GithubAgent(request.user.accessToken).getUser(name);;
    }
```

Testing is easy as i can use sinon to stub away the `GithubAgent` at unit test time.

I will look into the `AddUserToLocalsMiddleware ` solution to get rid of direct `@Req()` injection into my controller.
